### DISCUSSION TITLE
"Generator raised StopIteration" error when sending static file from filesystem

### DISCUSSION THREAD

**Discussion by ned-pcs** [miguelgrinberg/microdot#142] [Q&A]
I really like Microdot (which greatly resembles Sinatra, which I've used a lot in Ruby).

I'm using the latest version of Microdot-asyncio on an ESP32.
I would like to pre-compile my static files into frozen strings (and use `pkg_resources.resource_stream('static', 'filename')`).
I've tried both that method (which returns an `io.BytesIO` object), and serving these files from my filesystem.
I've tried returning an opened file handle in the `Response`, as well as using `send_file()`.
In all of the above cases, what I get is a `RuntimeError: generator raised StopIteration` exception:

```
Send file /lib/static/js/toast.js
Response header {'Content-Type': 'application/javascript', 'Cache-Control': 'max-age=86400'}
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "microdot/microdot_asyncio.py", line 269, in serve
  File "microdot/microdot_asyncio.py", line 337, in handle_request
  File "microdot/microdot_asyncio.py", line 155, in write
RuntimeError: generator raised StopIteration
```

Templates (using `utemplate`) and static strings work fine in the `Response`s.

Here's how I'm trying to return an opened file:
```
f = open(filename, 'rb')
return Response(body=f, status_code=status_code, headers=response_header)
```
Using `send_file()`:
```
return send_file(filename, status_code=status_code, content_type=content_type)
```
Sending from compiled resources:
```
f = pkg_resources.resource_stream('static', fname)
return Response(body=f, status_code=status_code, headers=response_header)
```
or sending from compiled resources using `send_file()`:
```
f = pkg_resources.resource_stream('static', fname)
return send_file(filename, status_code=status_code, content_type=content_type, stream=f)
```

Any suggestions?

**Comment by miguelgrinberg** [comment]
I believe this is a bug in the async version of Microdot's `send_file` function. I will look into it and provide a fix soon.

**Reply by ned-pcs** [reply]
I've also seen this error without using `send_file`, if that helps.

**Reply by miguelgrinberg** [reply]
How do you import the `send_file` function? Could you be importing it from `microdot` instead of `microdot_asyncio`? I've found that I've made a mistake in an import on the `static_async.py` example in this repository, so I'm wondering if maybe you based your code on it.

**Reply by ned-pcs** [reply]
I was importing it from `microdot_asyncio`. I've updated my code since then, and am no longer using `send_file`. It's possible I was running into another problem, though; as I've re-written chunks of the static-file sending code the issue has gone away.
Here's what I came up with for a generic static file serving routine, which so far is working for me (you're welcome to use it):

```python
import pkg_resources, os
from microdot_asyncio import Response

def path_exists(path):
    try:
        _ = os.stat(path)
        return True
    except OSError:
        return False

def response_for_file(filename: str, allow_gz: bool = False) -> Response:
    """
    Return the Response object for a static file.
    Static files are either served from the /lib/static/<filetype> directory,
    or from pre-compiled strings using pkg_resources in the 'static' namespace.
    Assumes static files are separated into directories based on their extension.

    :param      filename:  The filename
    :type       filename:  str

    :param      allow_gz:  Whether to allow gzipped files to be served.
    :type       allow_gz:  bool

    :returns:   The response for the file.
    :rtype:     Response
    """
    if '..' in filename:
        return Response(body='Not found', status_code=404, headers={}, reason='Directory traversal is not allowed')

    # split the filename into parts, and get the extension
    ext = filename.split('.')[-1]
    # get the content-type for the extension
    content_type = Response.types_map.get(ext, 'application/octet-stream')
    headers = {}
    headers['Content-Type'] = content_type
    headers['Cache-Control'] = 'max-age=86400'
    static_path = f"/lib/static/{ext}/{filename}"
    if allow_gz:
        zipped = static_path + '.gz'
        if path_exists(zipped):
            headers['Content-Encoding'] = 'gzip'
            return Response(body = open(zipped, 'rb'), status_code=200, headers=headers)
    elif path_exists(static_path):
        return Response(body = open(static_path, 'rb'), status_code=200, headers=headers)
    try:
        f = pkg_resources.resource_stream('static', f"{ext}/{filename}")
        return Response(body = f, status_code=200, headers=headers)
    except KeyError:
        return Response(body='File not found', status_code=404)

```
I am trying to freeze everything into flash, using `utemplate_util.py` for the `utemplate` templates, and using `mpy_bin2res.py` for the other static content.
